baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/IllegalSafeLoggingArgument.java

```diff
@@ -427,4 +432,76 @@ public Description matchClass(ClassTree tree, VisitorState state) {
                         "Dangerous type: annotated '%s' but ancestors declare '%s'.", directSafety, ancestorSafety))
                 .build();
     }
+
+    @Override
+    public Description matchLambdaExpression(LambdaExpressionTree tree, VisitorState state) {
+        Type variableType = ASTHelpers.getType(tree);
+
+        Type returnType = getFunctionalInterfaceReturnType(variableType, state);
+        if (returnType == null) {
+            return Description.NO_MATCH;
+        }
+        Safety requiredReturnSafety = SafetyAnnotations.getSafety(returnType, state);
+
+        if (requiredReturnSafety.allowsAll()) {
+            // Short-circuit if the return type allows all values
+            return Description.NO_MATCH;
+        }
+
+        Safety resultSafety = getLambdaReturnSafety(tree, state);
+
+        if (requiredReturnSafety.allowsValueWith(resultSafety)) {
+            return Description.NO_MATCH;
+        }
+
+        return buildDescription(tree)
+                .setMessage(String.format(
+                        "Dangerous return type in lambda function: expected return type is '%s' "
+                                + "but the lambda is returning '%s'.",
+                        requiredReturnSafety, resultSafety))
+                .build();
+    }
+
+    private Type getFunctionalInterfaceReturnType(Type variableType, VisitorState state) {
```

Comment:
Going to replace this entire method with `ASTHelpers.targetType(state.withPath(new TreePath(state.getPath(), tree)));` which delegates to `state.getTypes().findDescriptorType(getType(lambdaExpressionTree)).getReturnType();` and is a much simpler version of this